src/python/gem5/utils/looppoint.py

```diff
@@ -0,0 +1,404 @@
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from m5.util import fatal
+from m5.params import PcCountPair
+from pathlib import Path
+from typing import List, Dict
+from gem5.components.processors.abstract_processor import AbstractProcessor
+from m5.objects import PcCountTrackerManager
+import csv
+import re
+import json
+
+
+class LoopPoint:
+    """
+    This LoopPoint class is used to manage the information needed for LoopPoint
+    in workload
+
+    """
+
+    def __init__(
+        self,
+        targets: List[PcCountPair],
+        regions: Dict[PcCountPair, int],
+        json_file: Dict[int, Dict],
+    ) -> None:
+        """
+        :param targets: a list of PcCountPair that are used to generate exit
+        event at when the PcCountTrackerManager encounter this PcCountPair in
+        execution
+        :param regions: a dictionary used to find the corresponding region id
+        for the significant PcCountPair. This is mainly used to ensure
+        checkpoints are taken in the correct PcCountPair or relative counts are
+        updated at the correct count
+        :param json_file: all the LoopPoint data including relative counts and
+        multiplier are stored in this parameter. It can be outputted as a json
+        file.
+        """
+
+        self._manager = PcCountTrackerManager()
+        self._manager.targets = targets
+        self._targets = targets
+        self._regions = regions
+        self._json_file = json_file
+
+    def setup_processor(
+        self,
+        processor: AbstractProcessor,
+    ) -> None:
+        """
+        This function is used to setup a PC tracker in all the cores and
+        connect all the tracker to the PC tracker manager to perform
+        multithread PC tracking
+        :param processor: the processor used in the simulation configuration
+        """
+        for core in processor.get_cores():
+            core.add_pc_tracker_probe(self._targets, self._manager)
+
+    def update_relatives(self) -> None:
+        """
+        This function is used to update the relative count for restore used.
+        The new relative count will be stored in the _json_file and can be
+        outputted into a json file by calling the output_json_file function.
+        """
+        current_pair = self._manager.get_current_pc_count_pair()
+        temp_pair = PcCountPair(current_pair.getPC(), current_pair.getCount())
```

Comment:
Isn't the type of `current_pair` a `PcCountPair`? I don't understand why the `temp_pair` is needed.